iree/compiler/Dialect/HAL/Transforms/test/pack_allocations.mlir

```diff
@@ -0,0 +1,100 @@
+// RUN: iree-opt -split-input-file -iree-hal-target-backends=dylib-llvm-aot -iree-hal-pack-allocations -cse -canonicalize %s | IreeFileCheck %s
+
+// CHECK-LABEL: @packStatic
+// CHECK-SAME: %[[ALLOCATOR:.+]]: !hal.allocator
+func @packStatic(%allocator: !hal.allocator) ->
+    (index, index, index, index, index, index, index) {
+  %c100 = constant 100 : index
+  %c200 = constant 200 : index
+  %t:7 = hal.allocator.pack<%allocator : !hal.allocator> slices({
+    [0, 1] = %c100,  // +0
+    [1, 2] = %c100,  // +112 (100 align 16)
+    [2, 3] = %c100,  // +0 (reuse [0, 1])
+    [0, 4] = %c200,  // +224 (after 112 + 112; end align 16)
+    [5, 6] = %c200,  // +0 (reuse [0, 1]/[2, 3])
+    [5, 8] = %c100,  // +208 (after 200 align 16)
```

Comment:
this slice could reuse `[0, 1]` if `[5, 6]` was not already reusing it, right?

Reply:
Correct - and as you can imagine there's arrangements that have more or less fragmentation (if B was placed before A then there's no gap as there would be if A came before B, etc). There's tons of heuristics for this and in a lot of strip packing implementations you'll see things like "bottom left" or "bottom right" modes to try to accommodate for that ([like in stb](https://github.com/nothings/stb/blob/master/stb_rect_pack.h#L151-L161)). Some strategies will try several of these approaches and pick the best one as determining which heuristic to use can be as complex as just doing it and choosing the min :P